tools/offcputime.py

```diff
@@ -129,32 +130,46 @@ def signal_ignore(signal, frame):
     else:
         print("... Hit Ctrl-C to end.")
 
-# output
-while (1):
-    try:
-        sleep(duration)
-    except KeyboardInterrupt:
-        # as cleanup can take many seconds, trap Ctrl-C:
-        signal.signal(signal.SIGINT, signal_ignore)
-
-    if not folded:
-        print()
-    counts = b.get_table("counts")
-    stack_traces = b.get_table("stack_traces")
-    for k, v in sorted(counts.items(), key=lambda counts: counts[1].value):
-        if folded:
-            # print folded stack output
-            stack = list(stack_traces.walk(k.stack_id))[1:]
-            line = [k.name.decode()] + [b.ksym(addr) for addr in reversed(stack)]
-            print("%s %d" % (";".join(line), v.value))
-        else:
-            # print default multi-line stack output
-            for addr in stack_traces.walk(k.stack_id):
-                print("    %-16x %s" % (addr, b.ksym(addr)))
-            print("    %-16s %s" % ("-", k.name))
-            print("        %d\n" % v.value)
-    counts.clear()
-
-    if not folded:
-        print("Detaching...")
-    exit()
+try:
+    sleep(duration)
+except KeyboardInterrupt:
+    # as cleanup can take many seconds, trap Ctrl-C:
+    signal.signal(signal.SIGINT, signal_ignore)
+
+if not folded:
+    print()
+
+missing_stacks = 0
+counts = b.get_table("counts")
+stack_traces = b.get_table("stack_traces")
+for k, v in sorted(counts.items(), key=lambda counts: counts[1].value):
+    """
+    bpf_get_stackid will return a negative value in the case of an error
+
+    BPF_STACK_TRACE(_name, _size) will allocate space for _size stack traces
+    on each CPU. -ENOMEM will be returned when this limit is reached within a
```

Comment:
the last sentence is incorrect. The pcpu_freelist_pop() is a single free-list for all cpu that has 'per-cpu' optimization. When user space sees ENOMEM from get_stackid, it means that it reached the max_entries limit. For this tool 1024 is not enough then. Handling it gracefully makes sense as well.